Adding documentation on APIs no longer supported in Desktop (#4210)
microsoft/microsoft-ui-xaml: docs/winrt-apis-for-desktop.md

# WinRT API changes for Desktop apps

In WinUI 3 Preview 4, we have made changes to the way Windowing works in Desktop apps, by no longer creating a CoreWindow or CoreApplicationView. As a result, all WinRT APIs dependent on CoreWindow will no longer work in Desktop apps. This document will list out APIs that have been affected by this change, with replacement or workaround APIs that will allow you to achieve the same functionality. 

This doc will continue to be updated as our community files bugs and as we identify more workarounds/replacements. If you are encountering issues with an API not listed here, please go ahead and [file a bug](https://github.com/microsoft/microsoft-ui-xaml/issues/new?assignees=&labels=&template=bug_report.md&title=) on this repo with the API in question and what you are trying to achieve by using it. 

| WinRT API                                                                  | Replacement or Workaround                 | Notes                                             |
|----------------------------------------------------------------------------|-------------------------------------------|---------------------------------------------------|

> | [`ApplicationViewTitleBar.ExtendViewIntoTitleBar`](https://docs.microsoft.com/en-us/uwp/api/windows.applicationmodel.core.coreapplicationviewtitlebar.extendviewintotitlebar?view=winrt-19041)                            | `Window.ExtendsContentIntoTitleBar`        |  API reference documentation coming soon. This is a permanent replacement for WinUI 3.                                                |

Reviewer: I think this information is difficult to represent in a table. It's really ApplicationView that's not supported in Desktop apps, along with the dozen or so ApplicationView... classes and all their members, including ApplicationViewTitleBar.ExtendViewIntoTitleBar.

And it's all members of DisplayInformation that's not supported, not just the two called out here.

Maybe a better format would be to call out each major unsupported API (like ApplicationView or DisplayInformation), and then related classes, and alternatives for some of the members.

Author: @MikeHillberg This is great feedback - I've changed the structure in an upcoming commit, it would be great if you could take a look.

> | [`CoreWindow.GetForCurrentThread().GetKeyState()`]((https://docs.microsoft.com/en-us/uwp/api/windows.ui.core.corewindow.getkeystate?view=winrt-19041) )                             | [`KeyboardInput.GetKeyStateForCurrentThread`](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.input.keyboardinput.getkeystateforcurrentthread?view=winui-3.0-preview) |                                                   |

Reviewer: The GetForCurrentThread part isn't important, it's just that CoreWindow isn't supported.

| [`DisplayInformation.RawPixelsPerViewPixel`](https://docs.microsoft.com/uwp/api/windows.graphics.display.displayinformation.rawpixelsperviewpixel?view=winrt-19041) | [`XamlRoot.RasterizationScale`](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.xamlroot.rasterizationscale?view=winui-3.0-preview)               |  |
| [`Window.Dispatcher`](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.window.dispatcher?view=winui-3.0-preview) or [`CoreDispatcher`](https://docs.microsoft.com/en-us/uwp/api/windows.ui.core.coredispatcher?view=winrt-19041)                                               | [`DispatcherQueue`](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.window.dispatcherqueue?view=winui-3.0-preview)                           | DispatcherQueue is the permanent WinUI 3 replacement for Dispatcher and CoreDispatcher.

> | [`UIViewSettings.GetForCurrentView()`](https://docs.microsoft.com/en-us/uwp/api/windows.ui.viewmanagement.uiviewsettings.getforcurrentview?view=winrt-19041)  |                 | No alternative - this functionality is not intended to be supported in Desktop apps.  |

Reviewer: It's all of UIViewSettings that isn't supported on Desktop.

# WinRT API changes for Desktop apps

> Some Windows APIs are designed for and only completely supported in a UWP app, especially CoreDispatcher, CoreWindow, ApplicationView, and some related classes. These APIs are partially supported in Desktop apps when _hosting_ Xaml (Xaml Islands), and in WinUI3 previews before Preview4. As of WinUI3 Preview4 they are fully unsupported.

Reviewer 2: ```suggestion
Some Windows APIs are designed for and only completely supported in a UWP app, especially [CoreDispatcher](https://docs.microsoft.com/uwp/api/Windows.UI.Core.CoreDispatcher), [CoreWindow](https://docs.microsoft.com/uwp/api/Windows.UI.Core.CoreWindow), [ApplicationView](https://docs.microsoft.com/uwp/api/windows.ui.viewmanagement.applicationview), and some related classes. These APIs are partially supported in Desktop apps when _hosting_ XAML (XAML Islands), and in WinUI 3 previews before Preview 4. As of WinUI 3 Preview 4 they are fully unsupported.
```

> Note that when a class is unsupported in Desktop it can lead to other unsupported classes. For example, several classes in the Windows API have a static GetForCurrentView method, such as [UIViewSettings.GetForCurrentView](https://docs.microsoft.com/uwp/api/Windows.UI.ViewManagement.UIViewSettings.GetForCurrentView), which is a pattern that indicates a dependence on ApplicationView, and so are not supported in Desktop apps.

Reviewer 2: ```suggestion
Note that when an API is unsupported, other APIs that depend on the former become unsupported too. For example, several classes in the Windows Runtime API have a static `GetForCurrentView()` method, such as [UIViewSettings.GetForCurrentView](https://docs.microsoft.com/uwp/api/Windows.UI.ViewManagement.UIViewSettings.GetForCurrentView), which is a pattern that indicates a dependence on `ApplicationView`, and therefore these are not supported in Desktop apps.
```

This document lists out these APIs - these are no longer supported in Desktop apps, and neither are any of its members or other APIs that depend on it. There are suggested replacements below, where available, to achieve the same functionality. 

This doc is a live document and will continue to be updated as we identify more workarounds/replacements. If you are encountering issues with an API not listed here, please go ahead and [file a bug](https://github.com/microsoft/microsoft-ui-xaml/issues/new?assignees=&labels=&template=bug_report.md&title=) on this repo with the API in question and what you are trying to achieve by using it. 

## ApplicationView

> - [`ApplicationViewTitleBar.ExtendViewIntoTitleBar`](https://docs.microsoft.com/en-us/uwp/api/windows.applicationmodel.core.coreapplicationviewtitlebar.extendviewintotitlebar?view=winrt-19041) 

Reviewer 2: ```suggestion
- [`ApplicationViewTitleBar.ExtendViewIntoTitleBar`](https://docs.microsoft.com/uwp/api/windows.applicationmodel.core.coreapplicationviewtitlebar.extendviewintotitlebar) 
```

    
    Use `Window.ExtendsContentIntoTitleBar` instead. This API was added in Preview 4 and is intended as a permanent replacement in WinUI 3. API reference documentation for this is coming soon.

## CoreWindow

> - [`CoreWindow.GetKeyState()`](https://docs.microsoft.com/en-us/uwp/api/windows.ui.core.corewindow.getkeystate?view=winrt-19041) 

Reviewer 2: ```suggestion
- [`CoreWindow.GetKeyState()`](https://docs.microsoft.com/uwp/api/windows.ui.core.corewindow.getkeystate) 
```

>     Use [`KeyboardInput.GetKeyStateForCurrentThread`](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.input.keyboardinput.getkeystateforcurrentthread?view=winui-3.0-preview) instead.

Reviewer 2: ```suggestion
    Use [`KeyboardInput.GetKeyStateForCurrentThread`](https://docs.microsoft.com/windows/winui/api/microsoft.ui.input.keyboardinput.getkeystateforcurrentthread?view=winui-3.0-preview) instead.
```

> - [`CoreWindow.PointerCursor`](https://docs.microsoft.com/en-us/uwp/api/windows.ui.core.corewindow.pointercursor?view=winrt-19041) 
> 
>     Use [`UIElement.ProtectedCursor`](https://docs.microsoft.com/windows/winui/api/microsoft.ui.xaml.uielement.protectedcursor?view=winui-3.0-preview) instead. Note that you'll need to have a subclass of UIElement to access this API. 
> 
> ## CoreDispatcher
> Note too that the [Window.Dispatcher](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Window.Dispatcher) and [DependencyObject.Dispatcher](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.DependencyObject.Dispatcher) properties return null in a Desktop app.
> 
> - [`Window.Dispatcher`](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.window.dispatcher?view=winui-3.0-preview) or [`CoreDispatcher`](https://docs.microsoft.com/en-us/uwp/api/windows.ui.core.coredispatcher?view=winrt-19041)
> 
>     Use [`DispatcherQueue`](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.window.dispatcherqueue?view=winui-3.0-preview) instead. `DispatcherQueue` is the permanent WinUI 3 replacement for `Dispatcher` and `CoreDispatcher`.
> 
> ## CoreApplicationView
> This API is no longer supported in Desktop apps, and neither are any of its members or other APIs that depend on it.
> 
> ## DisplayInformation
> 
> - [`DisplayInformation.LogicalDpi`](https://docs.microsoft.com/en-us/uwp/api/windows.graphics.display.displayinformation.logicaldpi?view=winrt-19041)
>     
>     Use [`XamlRoot.RasterizationScale`](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.xamlroot.rasterizationscale?view=winui-3.0-preview) instead, and listen for changes on the [`XamlRoot.Changed`](https://docs.microsoft.com/uwp/api/windows.ui.xaml.xamlroot.changed?view=winrt-19041) event. 
> 
> - [`DisplayInformation.RawPixelsPerViewPixel`](https://docs.microsoft.com/uwp/api/windows.graphics.display.displayinformation.rawpixelsperviewpixel?view=winrt-19041)
> 
>     Use [`XamlRoot.RasterizationScale`](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.xamlroot.rasterizationscale?view=winui-3.0-preview) instead. 
> 
> ## UIViewSettings
> 
> - [`UIViewSettings.GetForCurrentView()`](https://docs.microsoft.com/en-us/uwp/api/windows.ui.viewmanagement.uiviewsettings.getforcurrentview?view=winrt-19041)
> 
>     No alternative - `UIViewSettings` is not intended to be supported in Desktop apps.
> 
> ## SystemNavigationManager
> 
> 
> - [`SystemNavigationManager.GetForCurrentView()`](https://docs.microsoft.com/en-us/uwp/api/windows.ui.core.systemnavigationmanager.getforcurrentview?view=winrt-19041)
> 
>     No alternative - this functionality is not intended to be supported in Desktop apps. 

Reviewer 2: ```suggestion
- [`CoreWindow.PointerCursor`](https://docs.microsoft.com/uwp/api/windows.ui.core.corewindow.pointercursor) 

    Use [`UIElement.ProtectedCursor`](https://docs.microsoft.com/windows/winui/api/microsoft.ui.xaml.uielement.protectedcursor?view=winui-3.0-preview) instead. Note that you'll need to have a subclass of `UIElement` to access this API. 

## CoreDispatcher
Note that the [Window.Dispatcher](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Window.Dispatcher) and [DependencyObject.Dispatcher](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.DependencyObject.Dispatcher) properties return `null` in a Desktop app.

- [`Window.Dispatcher`](https://docs.microsoft.com/windows/winui/api/microsoft.ui.xaml.window.dispatcher?view=winui-3.0-preview) or [`CoreDispatcher`](https://docs.microsoft.com/uwp/api/windows.ui.core.coredispatcher)

    Use [`DispatcherQueue`](https://docs.microsoft.com/windows/winui/api/microsoft.ui.xaml.window.dispatcherqueue?view=winui-3.0-preview) instead. `DispatcherQueue` is the permanent WinUI 3 replacement for `Dispatcher` and `CoreDispatcher`.

## CoreApplicationView
This API is no longer supported in Desktop apps, and neither are any of its members or other APIs that depend on it.

## DisplayInformation

- [`DisplayInformation.LogicalDpi`](https://docs.microsoft.com/uwp/api/windows.graphics.display.displayinformation.logicaldpi)

    Use [`XamlRoot.RasterizationScale`](https://docs.microsoft.com/windows/winui/api/microsoft.ui.xaml.xamlroot.rasterizationscale?view=winui-3.0-preview) instead, and listen for changes on the [`XamlRoot.Changed`](https://docs.microsoft.com/uwp/api/windows.ui.xaml.xamlroot.changed) event. 

- [`DisplayInformation.RawPixelsPerViewPixel`](https://docs.microsoft.com/uwp/api/windows.graphics.display.displayinformation.rawpixelsperviewpixel)

    Use [`XamlRoot.RasterizationScale`](https://docs.microsoft.com/windows/winui/api/microsoft.ui.xaml.xamlroot.rasterizationscale?view=winui-3.0-preview) instead. 

## UIViewSettings

- [`UIViewSettings.GetForCurrentView()`](https://docs.microsoft.com/uwp/api/windows.ui.viewmanagement.uiviewsettings.getforcurrentview)

    No alternative - `UIViewSettings` is not intended to be supported in Desktop apps.

## SystemNavigationManager

- [`SystemNavigationManager.GetForCurrentView()`](https://docs.microsoft.com/uwp/api/windows.ui.core.systemnavigationmanager.getforcurrentview)

    No alternative - this functionality is not intended to be supported in Desktop apps. 
```

# WinRT API changes for Desktop apps

> In WinUI 3 Preview 4, we have made changes to the way Windowing works in Desktop apps, by no longer creating a CoreWindow or CoreApplicationView. As a result, many WinRT APIs will no longer be supported in Dekstop apps. This document will list out APIs that have been affected by this change, with replacement or workaround APIs that will allow you to achieve the same functionality. 

Reviewer: Suggesting some more background, and pulling to the top that note that APIs that depend on unsupported classes also become unsupported.

```suggestion
Some Windows APIs are designed for and only completely supported in a UWP app, especially CoreDispatcher, CoreWindow, ApplicationView, and some related classes. These APIs are partially supported in Desktop apps when _hosting_ Xaml (Xaml Islands), and in WinUI3 previews before Preview4. As of WinUI3 Preview4 they are fully unsupported.

Note that when a class is unsupported in Desktop it can lead to other unsupported classes. For example, several classes in the Windows API have a static GetForCurrentView method, such as [UIViewSettings.GetForCurrentView](https://docs.microsoft.com/uwp/api/Windows.UI.ViewManagement.UIViewSettings.GetForCurrentView), which is a pattern that indicates a dependence on ApplicationView, and so are not supported in Desktop apps.

This document lists out these APIs, and suggests replacements where available to achieve the same functionality. 
```

This doc will continue to be updated as our community files bugs and as we identify more workarounds/replacements. If you are encountering issues with an API not listed here, please go ahead and [file a bug](https://github.com/microsoft/microsoft-ui-xaml/issues/new?assignees=&labels=&template=bug_report.md&title=) on this repo with the API in question and what you are trying to achieve by using it. 

## ApplicationView

> This API is no longer supported in Desktop apps, and neither are any of its members or other APIs that depend on it. Below is a list of workarounds the team has identified for members or dependencies of the ApplicationView class.

Reviewer: I think we can drop this paragraph from each of the sections, but covering it at the top?

Author: Done

    
    Use `Window.ExtendsContentIntoTitleBar` instead. This API was added in Preview 4 and is intended as a permanent replacement in WinUI 3. API reference documentation for this is coming soon.

## CoreWindow
This API is no longer supported in Desktop apps, and neither are any of its members or other APIs that depend on it. Below is a list of workarounds the team has identified for members or dependencies of the CoreWindow class.

- [`CoreWindow.GetKeyState()`]((https://docs.microsoft.com/en-us/uwp/api/windows.ui.core.corewindow.getkeystate?view=winrt-19041) ) 

    Use [`KeyboardInput.GetKeyStateForCurrentThread`](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.input.keyboardinput.getkeystateforcurrentthread?view=winui-3.0-preview) instead.

- [`CoreWindow.PointerCursor`](https://docs.microsoft.com/en-us/uwp/api/windows.ui.core.corewindow.pointercursor?view=winrt-19041) 

    Use [`UIElement.ProtectedCursor`](https://docs.microsoft.com/windows/winui/api/microsoft.ui.xaml.uielement.protectedcursor?view=winui-3.0-preview) instead. Note that you'll need to have a subclass of UIElement to access this API. 

## CoreDispatcher

> This API is no longer supported in Desktop apps, and neither are any of its members or other APIs that depend on it. Below is a list of workarounds the team has identified for members or dependencies of the CoreDispatcher class.

Reviewer: ```suggestion
Note too that the [Window.Dispatcher](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Window.Dispatcher) and [DependencyObject.Dispatcher](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.DependencyObject.Dispatcher) properties return null in a Desktop app.
```

This document lists out these APIs - these are no longer supported in Desktop apps, and neither are any of its members or other APIs that depend on it. There are suggested replacements below, where available, to achieve the same functionality. 

This doc is a live document and will continue to be updated as we identify more workarounds/replacements. If you are encountering issues with an API not listed here, please go ahead and [file a bug](https://github.com/microsoft/microsoft-ui-xaml/issues/new?assignees=&labels=&template=bug_report.md&title=) on this repo with the API in question and what you are trying to achieve by using it. 

## ApplicationView

- [`ApplicationViewTitleBar.ExtendViewIntoTitleBar`](https://docs.microsoft.com/uwp/api/windows.applicationmodel.core.coreapplicationviewtitlebar.extendviewintotitlebar) 
    
    Use `Window.ExtendsContentIntoTitleBar` instead. This API was added in Preview 4 and is intended as a permanent replacement in WinUI 3. API reference documentation for this is coming soon.
    

> As mentioned above, `GetForCurrentView()` methods have a dependence on `ApplicationView` and are not supported in Desktop apps. Note that some `GetForCurrentView` methods will not only return null, but will also throw exceptions. We advise checking if `Window.Current()` is null before calling these APIs - although `Window.Current()` returning null may be expected in some scenarios, such as during background activation. 

Reviewer 2: it's either `Window.Current` (no parens) in C# or `Window::Current()` in cppwinrt :)

Author: Ah just missed this before merging... fixing now :)

Reviewer 2: ```suggestion
As mentioned above, `GetForCurrentView()` methods have a dependence on `ApplicationView` and are not supported in Desktop apps. Note that some `GetForCurrentView` methods will not only return null, but will also throw exceptions. We advise checking if `Window.Current` is null before calling these APIs - although `Window.Current` returning null may be expected in some scenarios, such as during background activation. 
```
